connectors/memory/memory.go

```diff
@@ -0,0 +1,493 @@
+	// no data in this partition? easy out!
+	if len(partitionRef) == 0 {
+		entityRef[encodedPartitionKey] = append(entityRef[encodedPartitionKey], values)
+		return nil
+	}
+
+	if len(ei.Def.ClusteringKeySet()) == 0 {
+		// no clustering key, so the row must already exist, merge it
+		merge(partitionRef[0], values)
+		return nil
+	}
+	// there is a clustering key, find the insertion point (binary search would be fastest)
+	found, offset := findInsertionPoint(ei, partitionRef, values)
+	if found {
+		merge(partitionRef[offset], values)
+		return nil
+	}
+	// perform slice magic to insert value at given offset
+	l := len(entityRef[encodedPartitionKey])                                                                     // get length
+	entityRef[encodedPartitionKey] = append(entityRef[encodedPartitionKey], entityRef[encodedPartitionKey][l-1]) // copy last element
+	// scoot over remaining elements
+	copy(entityRef[encodedPartitionKey][offset+1:], entityRef[encodedPartitionKey][offset:])
+	// and plunk value into appropriate location
+	entityRef[encodedPartitionKey][offset] = values
+	return nil
+}
+func merge(into map[string]dosa.FieldValue, from map[string]dosa.FieldValue) {
+	for k, v := range from {
+		into[k] = v
+	}
+}
+
+// Remove deletes a single row
+func (c *Connector) Remove(_ context.Context, ei *dosa.EntityInfo, values map[string]dosa.FieldValue) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.data[ei.Def.Name] == nil {
+		return &dosa.ErrNotFound{}
+	}
+	entityRef := c.data[ei.Def.Name]
+	encodedPartitionKey := partitionKeyBuilder(ei, values)
+	if entityRef[encodedPartitionKey] == nil {
+		return &dosa.ErrNotFound{}
+	}
+	partitionRef := entityRef[encodedPartitionKey]
+	// no data in this partition? easy out!
+	if len(partitionRef) == 0 {
+		return &dosa.ErrNotFound{}
+	}
+
+	// no clustering keys? Simple, delete this
+	if len(ei.Def.ClusteringKeySet()) == 0 {
+		entityRef[encodedPartitionKey] = nil
+		return nil
+	}
+	found, offset := findInsertionPoint(ei, partitionRef, values)
+	if found {
+		entityRef[encodedPartitionKey] = append(entityRef[encodedPartitionKey][:offset], entityRef[encodedPartitionKey][offset+1:]...)
+		return nil
+	}
+	return &dosa.ErrNotFound{}
+}
+
+// Range returns a slice of data from the datastore
+func (c *Connector) Range(_ context.Context, ei *dosa.EntityInfo, columnConditions map[string][]*dosa.Condition, fieldsToRead []string, token string, limit int) ([]map[string]dosa.FieldValue, string, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	if c.data[ei.Def.Name] == nil {
+		return nil, "", &dosa.ErrNotFound{}
+	}
+	entityRef := c.data[ei.Def.Name]
+
+	// find the equals conditions on each of the partition keys
+	values := map[string]dosa.FieldValue{}
+	for _, pk := range ei.Def.Key.PartitionKeys {
+		// TODO: assert len(columnConditions[pk] == 1
+		// TODO: assert columnConditions[pk][0].Op is equals
+		values[pk] = columnConditions[pk][0].Value
+	}
+
```

Comment:
Connectors can safely assume their inputs have been sanitized, per our early design discussions. This is because we are controlling the callers. If we want to have a sanitizing connector, we can do that too, but adding lots of checks at every layer isn't necessary.